### ISSUE TITLE
Addition of comment export functionality (per video, as JSON, format example see in comment).

### ISSUE THREAD

**Issue by preai** [mattwright324/youtube-comment-suite#13]
```
[
  {
    "id": "UgzXyVGsIiMcZ9Bs-M54AaABAg",
    "user": "abc",
    "date": "4 days ago",
    "timestamp": 1558387486139,
    "commentText": "bla bla bla",
    "likes": 1,
    "hasReplies": true,
    "numberOfReplies": 2,
    "replies": [
      {
        "id": "UgzXyVGsIiMcZ9Bs-M54AaABAg.8vAG3f1HbIR8vARqlHhxyM",
        "user": "def",
        "date": "3 days ago",
        "timestamp": 1558473886511,
        "commentText": "@abc     ha ha ha",
        "likes": 0
      },
      {
        "id": "UgzXyVGsIiMcZ9Bs-M54AaABAg.8vAG3f1HbIR8vBliOzYonT",
        "user": "ghi",
        "date": "3 days ago",
        "timestamp": 1558473886512,
        "commentText": "@def --- no no no!.",
        "likes": 0
      }
    ]
  }
]
```

**Comment by preai** [comment]
As implemented here:

http://ytcomments.klostermann.ca/

**Comment by mattwright324** [comment]
I can see the value in having this functionality and is similarly something I have in my [youtube-geofind](https://mattw.io/youtube-geofind/location) web tool. This should be fairly simple to implement. I'll definitely look into adding this somehow.

**Comment by mattwright324** [comment]
I'm fairly close to finishing this though there are some caveats about the right behavior and wonder what you were expecting. I have two export modes, flattened (comments and replies on the same level like they appear in the app) and condensed (how they would appear from that example/site and on the YouTube site, where replies are in a list as part of a parent comment). 

Flattened mode is great and easy haha.

Searching for "Comments and Replies" in YCS with default/no filters is how that site works and exporting results in condensed mode should result in almost the same thing.

However, you can't have condensed mode when you are searching "Comments Only" and "Replies Only", they would always have to be flattened. You could only export in condensed mode when searching both "Comments and Replies" because you kinda need both haha.

There is also the situation of a filtered query
- A reply when the parent comment filtered out, leave at base level or omit from export file?
- A parent comment with some or all replies filtered out, only show filtered ones if they fit the query? Show all replies regardless of query filters (similar to how 'view thread' works in app)?

**Comment by preai** [comment]
Hey Matthew,

great, thanks so much for your efforts! If it's too much work to integrate
the condensed mode, just stay with flattened, no problem. As long as you
have all comments in the export it shouldn't matter too much. I did notice
that on YouTube comments get omitted if you don't sort by newest, if those
end up in the export that would be great of course. I don't really care
about filtering for export, I just want all comments/replies in there, so I
can save them next to the video file. Sure it's n2h, but don't kill too
many of your brain cells with it. :D Just do it however you feel is most
beneficial without being too complicated to implement. ;-)

**Comment by mattwright324** [comment]
What if I we're to only allow condensed mode in ideal conditions which would satisfy what you've described is what you want?
- Only condensed mode is available when "Comments and Replies" is used and Name Like & Text Like are empty?
- Otherwise, flattened mode.

I'll note that currently condensed mode is also slower than flattened as it has to process comment threads (grab all comments in the thread) as they appear in the results list per video.

The folder structure when exporting will look like the following:
    
    folder-where-ycs-is-located/
        youtube-comment-suite.jar
        {other files...}
        exports/
            {yyyy.MM.dd hh.mm.ss}/
                searchSettings.json
                videoId-meta.json
                videoId-comments.json

Following this structure, the generated export folder `{yyyy.MM.dd hh.mm.ss}/` will have the date and time that the export button was pressed e.g. `2019.06.26 19.44.12/`. The file `searchSettings.json` will give some context for what your Search Comments tab options were when exporing. All comments will be exported per-video as you've suggested in the format of `videoId-meta.json` describing the details of the video (id, title, description, published date, etc.) and `videoId-comments` which will be an array of all comments either in the condensed or flattened format.

**Comment by Flashtrax** [comment]
Sorry if I missed something; but could the channel ID also be included (which left the comment) along with the channel ID which uploaded the video please?

**Comment by mattwright324** [comment]
Yup, those fields are included per `{videoId}-meta.json` and per comment in each `{videoId}-comments.json` file.

**Comment by preai** [comment]
Sounds awesome!

**Comment by mattwright324** [comment]
This has been mostly implemented in 073be22667faee3626a4788c6ab089ca04c378fd on the replace-ytapi branch. Behaves as described in previous comments. There is still some testing and tweaking to do but it will be in the next release.

**Comment by DontBLB** [comment]
Is it possible for there to be a function to export all comments as a CSV or excel please?
